src/test/java/org/prebid/server/analytics/reporter/greenbids/GreenbidsAnalyticsReporterTest.java

```diff
@@ -0,0 +1,643 @@
+        when(httpClient.post(anyString(), any(MultiMap.class), anyString(), anyLong()))
+                .thenReturn(Future.succeededFuture(mockResponse));
+
+        // when
+        final Future<Void> result = target.processEvent(event);
+
+        // then
+        assertThat(result.failed()).isTrue();
+        assertThat(result.cause())
+                .hasMessageStartingWith("Unexpected response status: 500");
+    }
+
+    @Test
+    public void shouldFailWhenAdUnitsListIsEmpty() {
+        // given
+        final AuctionContext auctionContext = mock(AuctionContext.class);
+        final BidResponse bidResponse = mock(BidResponse.class);
+        when(auctionContext.getBidRequest())
+                .thenReturn(BidRequest.builder()
+                                .id("request1")
+                                .ext(givenExtRequest())
+                                .build());
+
+        final AuctionEvent event = mock(AuctionEvent.class);
+        when(event.getAuctionContext()).thenReturn(auctionContext);
+        when(event.getBidResponse()).thenReturn(bidResponse);
+
+        // when
+        final Future<Void> result = target.processEvent(event);
+
+        // then
+        assertThat(result.failed()).isTrue();
+        assertThat(result.cause())
+                .hasMessageStartingWith("Impressions list should not be empty");
+    }
+
+    @Test
+    public void shouldFailOnDecodingImpExtPrebid() {
+        // given
+        final String prebid = "prebid";
+        final String bidderKey = "bidder";
+        final String optionKey = "options";
+        final String bidderName = "bidderName";
+        final String optionValue = "1";
+
+        final Map<String, Map<String, String>> bidderValue =
+                singletonMap(
+                        bidderName,
+                        Map.of("test-host", "unknownHost", "publisher_id", "ps4"));
+
+        final ObjectNode prebidJsonNodes = mapper.valueToTree(
+                singletonMap(
+                        prebid,
+                        Map.of(optionKey, optionValue, bidderKey, bidderValue)));
+
+        final Imp imp = Imp.builder()
+                .ext(prebidJsonNodes)
+                .banner(givenBanner())
+                .build();
+        final AuctionContext auctionContext = givenAuctionContext(imp);
+        final AuctionEvent event = AuctionEvent.builder()
+                .auctionContext(auctionContext)
+                .bidResponse(auctionContext.getBidResponse())
+                .build();
+
+        // when
+        final Future<Void> result = target.processEvent(event);
+
+        //then
+        assertTrue(result.failed());
+        assertThat(result.cause())
+                .hasMessageStartingWith("Error decoding imp.ext.prebid: "
+                        + "Cannot construct instance of `org.prebid.server.proto.openrtb.ext.request.ExtOptions`");
+    }
+
+    private static AuctionContext givenAuctionContext(Imp imp) {
```

Comment:
Again customizers. There should be only one givenAuctionContext but with customizers as method parameters. Examples can be found in bidder tests.